ark/phenotyping/som_utils.py

```diff
@@ -63,27 +64,107 @@ def compute_cluster_avg(fovs, channels, base_dir,
                             compression='uncompressed')
 
 
-def create_pixel_matrix(img_xr, seg_labels, base_dir,
+def create_fov_pixel_data(fov, channels, img_data, seg_labels, base_dir,
+                          pre_dir='pixel_mat_preprocessed', sub_dir='pixel_mat_subsetted',
+                          blur_factor=2, subset_proportion=0.1, seed=42):
+    """Preprocess pixel data for one fov
+
+    Saves preprocessed data to pre_dir and subsetted data to sub_dir
+
+    Args:
+        fov (str):
+            Name of the fov to index
+        channels (list):
+            List of channels to subset over
+        img_data (numpy.ndarray):
+            Array representing image data for one fov
+        seg_labels (numpy.ndarray):
+            Array representing segmentation labels for each image
+        base_dir (str):
+            Name of the directory to save the pixel files to
+        pre_dir (str):
+            Name of the directory which contains the preprocessed pixel data
+        sub_dir (str):
+            The name of the directory containing the subsetted pixel data
+        blur_factor (int):
+            The sigma to set for the Gaussian blur
+        subset_proportion (float):
+            The proportion of pixels to take from each fov
+        seed (int):
+            The random seed to set for subsetting
+    """
+
+    # for each marker, compute the Gaussian blur
+    for marker in range(len(channels)):
+        img_data[:, :, marker] = ndimage.gaussian_filter(img_data[:, :, marker],
+                                                         sigma=blur_factor)
+
+    # flatten each image
+    pixel_mat = img_data.reshape(-1, len(channels))
+
+    # convert into a dataframe
+    pixel_mat = pd.DataFrame(pixel_mat, columns=channels)
+
+    # assign metadata about each entry
+    pixel_mat['fov'] = fov
+    pixel_mat['row_index'] = np.repeat(range(img_data.shape[0]), img_data.shape[1])
+    pixel_mat['column_index'] = np.tile(range(img_data.shape[0]), img_data.shape[1])
+
+    # assign segmentation label
+    # seg_labels_flat = seg_labels.loc[fov, ...].values.flatten()
+    seg_labels_flat = seg_labels.flatten()
+    pixel_mat['segmentation_label'] = seg_labels_flat
+
+    # remove any rows that sum to 0
+    pixel_mat = pixel_mat.loc[pixel_mat.loc[:, channels].sum(axis=1) != 0, :]
+
+    # normalize each row by total marker counts to convert into frequencies
+    pixel_mat.loc[:, channels] = pixel_mat.loc[:, channels].div(
+        pixel_mat.loc[:, channels].sum(axis=1), axis=0)
+
+    # subset the pixel matrix for training
+    pixel_mat_subset = pixel_mat.sample(frac=subset_proportion, random_state=seed)
+
+    # write complete dataset to feather, needed for cluster assignment
+    feather.write_dataframe(pixel_mat,
```

Comment:
If we have this function return the pixel_mat and pixel_mat_subset, rather than save to disk, we don't have to pass in the paths to the directories. Then the top level function can do the saving

Reply:
Good idea.